e2e/cypress/tests/integration/performance/team_switch_spec.js

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+// ***************************************************************
+// - [#] indicates a test step (e.g. # Go to a page)
+// - [*] indicates an assertion (e.g. * Check the title)
+// - Use element ID when selecting an element. Create one if none.
+// ***************************************************************
+
+// Stage: @prod
+// Group: @channel
+import {measure} from './utils.js'
+
+describe('Channel switch performance test', () => {
+    let testUser;
+    let testTeam2;
+    before(() => {
+        cy.apiInitSetup().then(({team, user}) => {
+            testUser = user;
+
+            // # Login as test user and go to town square
+            cy.apiLogin(testUser);
+            cy.visit(`/${team.name}/channels/town-square`);
+        });
+
+        cy.apiCreateTeam('team-b', 'Team B').then(({team}) => {
+            testTeam2 = team;
+        });
+    });
+
+    it('measures switching between two teams from LHS', async () => {
+        // # Invoke window object
+
+        await measure('teamLoad', 500, () => {
```

Comment:
There is an issue with `measure`: Since measure runs multiple times, after the first run the user will already be on the page (e.g. offTopic), so for runs 2 and beyond, there is no switching between channels happening.

One way to solve this in this specific case is for the callback passed to measure (this function here) to switch between the two channels (so to switch from town-square to off-topic, then from off-topic to town-square etc (ie having state)).

The other way would be to have a reset callback passed to measure that would reset to the original state of test (ie go to the town-square channel in this case). This is trickier though, as you would need to exclude the duration of the reset from `measure`. E.g. something like that:

```javascript
export const measure = async (name, upperLimit, cb, reset, runs=10) => {
    return cy.window().its('performance').then(async (performance) => {
        const durations = [];

        for(let i = 0; i < runs; i++) {
            await reset();
            const perfName = `${name}_run_${i}`;

            performance.mark(perfName);
            await cb();
            performance.measure(perfName);

            const measure = performance.getEntriesByName(name)[0];
            const durationMs = Math.round(measure.duration/runs);
            durations.push(durationMs);
        }
    
        const avgDuration = Math.round(durations.reduce((a, b) => a + b)) / runs);
        cy.log(`[PERFORMANCE] ${name}: ${avgDuration}ms`, avgDuration <= upperLimit ? ' within upper limit' : ' outside upper limit', ` of ${upperLimit}ms`);
    
        assert.isAtMost(avgDuration, upperLimit);    
    });
};
```
The example above while more complex has the advantage that it could potentially report the individual run results too (e.g. log from each loop/run to have eventually a distribution of the durations).

Reply:
my bad, locally when running the tests it seems that it was measuring the switch itself a bunch of times, since this is what we are benchmarking. Will wait on the feedback from Saturn however to see if there is a reason to not allow async functions in the tests.

Reply:
This is important, because if we want to do the proper statistics behind the check, we need not only the average, but also the variance.

Reply:
Are we thinking of having every check run 10 times per test? Just clearing things out. I was initially thinking of even having one to three runs per PR as this would seem to suffice in other cypress performance tests example. IE, checking for page load is usually done once.

> One way to solve this in this specific case is for the callback passed to measure (this function here) to switch between the two channels (so to switch from town-square to off-topic, then from off-topic to town-square etc (ie having state)).

Nit: best practices recommend against storing state. 

Regarding the page load test, I guess we can get all load metrics by using the lighthouse API, example here: 
https://applitools.com/blog/using-cypress-google-lighthouse-performance-testing/